### ISSUE TITLE
[FEATURE] Allow other modules to hide an item from the HUD + make system ActionHandler part of api (for Items with Spells integration)

### ISSUE THREAD

**Issue by morepurplemorebetter** [Larkinabout/fvtt-token-action-hud-dnd5e#95]
A long title, but hopefully it conveys well what I'm asking. Yes, these are two separate things, but both are needed for getting my module Items with Spells to work with TAH. (Yes that module is mine since about 4 hours ago, why do you ask?)

I'm opening this issue more to have a discussion if you would be open to this idea and how you would prefer it be implemented.

## Describe your idea
The Items with Spells (IWS) module basically imports spells along when adding an item to an actor. It then sets some values on that spell depending on the setting on the item and links the spell to the item. The module does two things for displaying the spells:
1. It makes a custom section on the spellbook for each item with their spells therein, and
2. it show/hides this section depending on the item being attuned/equipped/identified.

These two things don't match with Token Action HUD yet, so I propose some changes to your module.

### 1. Custom sections
This is already possible, and I got it working by following the example for the Magic Items module (and doing a lot of reading of the base code of both DnD5e and Core modules). Thank you for making this possible!

However, I found it a bit underwhelming that I had just spent effort on fixing up the styling of the actions in this module, only to find that I couldn't call those same functions when it came to integrating my module.

I propose two changes to remedy this:
-  Expose the extended `ActionHandler` from the TAH system module as part of its `api`.
-  Change the function `#buildActions` to a public function `buildActions` so that it can be called though the new `api`.

### 2. Exclude items
If another module is adding items to a new group, it would be a shame if duplicates of those items appear in the HUD.
And to take an example from IWS, it would also be really great if the spells could be hidden when the item is also not visible.

I don't know how this would best be accomplished.
A flag on an entity? A hook that is called for every entity processed by TAH?

Ideally, this would have two options that TAH follows:
- An option where the entity is not shown in any of the TAH default groups (e.g. unidentified item, so don't show any of its ~~spoilers~~ spells to the player).
- An option where the entity is not shown in the groups visible by default (e.g. "1st-level spells"), but is visible in the activation groups (e.g. "Reactions"). This would be useful for IWS to set for every spell imported by it, so they don't show up in their regular group of At-will spells, but are available in other groups they would logically belong to.

## Pull request
I'm happy to make a working version of this and submit a PR for review. However, for that I would like your opinion on how best to tackle "2. Exclude items".
I think a hook would be great, but there would be a lot of calls to it, so I'm not sure if it would be too much overhead.

**Comment by Larkinabout** [comment]
Opening up the `buildActions` method to other modules sounds like an easy yes, though we may need to consider making it more versatile, maybe allowing some options to be passed in where needed.

For the second, I'd considered splitting the `addActions`  method to allow all actions to be built upfront, a single hook called to allow modification of any actions (such as removing them if they should be hidden from the player), then a method to add the actions to the HUD. This would allow other modules to intercept actions before they are added to the HUD but all in one go rather than piecemeal. This should be possible to do on the Core side without breaking the public methods.

It's important to consider that a user may also actually want duplicates in the HUD, for example, if they has two different types of categorisation for different circumstances.

**Comment by morepurplemorebetter** [comment]
Contrary to my previous post it seems it is not necessary to expose `ActionHandler` in the system module's api. Using the hook `tokenActionHudCoreAddActionHandlerExtenders`, I can get it to work just fine without. Just need to make `buildActions` public. Making that a bit more flexible could be useful, but my use-case doesn't have any options that would be useful to pass to it.

Perhaps put the settings in an object that can be overridden with passed options? I mean stuff like if spell details should be shown and things like that.

Your solution for the second point sounds like a great way to do it. Although I can't really see it yet, as there are multiple calls to `addActions` from the system module. But I trust you know better than me ;)

&nbsp;

> It's important to consider that a user may also actually want duplicates in the HUD, for example, if they has two different types of categorisation for different circumstances.

I completely agree and that's why I suggested having a way for the entity to only be excluded from those sections that are visible by default (e.g. Weapons) while not being excluded from those that can optionally be added (e.g. Bonus Actions).

However, sometimes one does want to exclude en entity from everything because that entity should be hidden.

**Comment by Larkinabout** [comment]
> Perhaps put the settings in an object that can be overridden with passed options? I mean stuff like if spell details should be shown and things like that.

Yes, allow for an object to be passed with some options. Just thinking ahead, so if it comes up in the future, it won't introduce a breaking change for other modules.

> Your solution for the second point sounds like a great way to do it. Although I can't really see it yet, as there are multiple calls to `addActions` from the system module. But I trust you know better than me ;)

Idea is a bit rough but, I'm thinking something like the existing `addActions` would add the actions into `availableActionsMap` as it does now but along with the group data, then here: https://github.com/Larkinabout/fvtt-token-action-hud-core/blob/d120f78cbfb35b3c67f30418bcbdc679edcb4962/scripts/action-handlers/action-handler.js#L106 - add in a method to call a hook and then add the actions to the HUD.

I'll see if I can start looking into it once I get all other modules sorted for V12.

**Comment by morepurplemorebetter** [comment]
Awesome!

Until that time, would it be okay with you if I publish a release on my TAH DnD5e fork with a quick and dirty fix for Items with Spells? Something like version `1.5.6.1` or version `1.5.6+iws`?
I'll remove it when we have a working integration.

The only changes would be:
- `#buildActions` to `buildActions`
- Add `if` statement in `#buildSpells` to skip over any spell gained from Items with Spells

**Comment by Larkinabout** [comment]
I've made `buildActions` and `buildActivations` available via `game.tokenActionHud.actionHandler`. Might add them to `game.modules...api` in the future, but the former should always be available also.

Feel free to do whatever you need to before I work on TAH Core to allow actions to be removed as part of the HUD build.

**Comment by morepurplemorebetter** [comment]
> Feel free to do whatever you need to before I work on TAH Core to allow actions to be removed as part of the HUD build.

Thank you. I made a release on my fork: [v1.5.7+iws](https://github.com/morepurplemorebetter/fvtt-token-action-hud-dnd5e/releases/tag/v1.5.7%2Biws).

It actually also fixes two other very minor issues in action-handler, see [the compare](https://github.com/morepurplemorebetter/fvtt-token-action-hud-dnd5e/compare/main...v1.5.7+iws).
